## Preparing Notebook

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Matt\Documents\Data Analysis Course\02-07-2022 InstaCart Basket Analysis'

In [3]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

## Mixed-Type Columns

In [7]:
# Create a Dataframe
df_test = pd.DataFrame()

In [8]:
# Create a Mixed-Type Column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# Correcting data type
df_test ['mix'] = df_test ['mix'].astype('str')

## Missing Values

In [16]:
# Checking for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
# Creating dataframe of missing values
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
# Checking shape of df_prods
df_prods.shape

(49693, 5)

In [20]:
# Creating new dataframe with removed missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [21]:
df_prods_clean.shape

(49677, 5)

## Duplicate Values

In [22]:
# Finding full duplicate values
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
# Checking shape of df_prods_clean
df_prods_clean.shape

(49677, 5)

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [27]:
df_prods_clean_no_dups.shape

(49672, 5)

In [29]:
# Exporting cleaned dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Task

## Importing Wrangled orders data

In [30]:
df_ords_wrangled = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## Question 2

In [31]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### The max value for prices must be a typo. With the median being 7.10 and the mean being 9.99, the max value of 99999.00 makes no sense.

## Question 3

In [32]:
# Checking for mixed data types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### Since 'mix' didn't come back in the output, that should mean there are my mixed type columns in df_ords.

## Question 5

In [33]:
# Checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### The only column with missing values is the Day Since Prior Order column. The only reason I can think of for the missing values is user error.

## Question 6

In [35]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [42]:
# Imputing using the mean
df_ords['days_since_prior_order'].fillna(1.111484e+01, inplace = True)

In [43]:
# Checking to see if the missing values are gone
df_ords.isnull().sum()

order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

#### I used mean imputation because removing 206,209 values from the data frame seemed excessive, and since the mean is closer to the minimum value (0) than the max value (30), I assumed most values are closer to the mean rather than the median.

## Question 7

In [44]:
df_dups_2 = df_ords[df_ords.duplicated()]

In [45]:
df_dups_2

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#### There doesn't seem to be any duplicate values in df_ords.

## Question 9

In [47]:
# Exporting Data
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))